### ISSUE TITLE
ApolloParseException caused by check failed when invoking guessNumber

### ISSUE THREAD

**Issue by rev0lgenX** [apollographql/apollo-kotlin#3836]
**Summary**
When mapping fromJson to generated class, in `JsonReaders#guessNumber` if `nextInt(MapJsonReader)` throws exception then rest of the method inside `guessNumber` fails and `nextNumber` will throw exception causing ApolloParseException .

**Version**
3.0.0

**Description**
When `nextInt` from `MapJsonReader` is invoked from `JsonReaders#guessNumber,` then  `MapJsonReader#nextValue` is invoked which will make `pendingName = null` in case of OrderedMap. Then, if nextInt throws error then rest of method call inside `JsonReaders#guessNumber` will fail because pendingName is null so when nextNumber is invoked, it invokes `MapJsonReader#nextValue` which will do `check(pendingName != null)` and throws exception.

![1](https://user-images.githubusercontent.com/20517801/152172262-618b4f90-6bee-4a8b-a565-dee0145c24d7.PNG)

![2](https://user-images.githubusercontent.com/20517801/152172288-4101a97b-8ddb-4ece-8324-87a775013257.PNG)

HELP!!!

**Comment by rev0lgenX** [comment]
Hi, 
Here is a sample that demonstrate the issue. [Sample](https://github.com/rev0lgenX/apollo-kotlin-sample)

Parse exception only occurs when fragment is used. Here is the query and it's response

## Query
```
query MediaListQuery($userId:Int,$mediaId:Int $type:MediaType){

    mediaList: MediaList(mediaId: $mediaId, userId: $userId, type: $type){
        ... mediaListEntry

    }
}


fragment mediaListEntry on MediaList {
    advancedScores
    media{
        title{
            english
        }
    }
}
```

## Response

```
{
  "data": {
    "MediaList": {
      "advancedScores": {
        "Story": 9.5,
        "Characters": 0,
        "Audio": 0,
        "Enjoyment": 0
      },
      "media": {
        "title": {
          "english": "The Rising of the Shield Hero"
        }
      }
    }
  }
}
```

#### Issue is with the advancedScores in fragment which have double as value. Exception is not thrown when I don't use fragment.

**Comment by BoD** [comment]
Thanks a lot for the repro project, it helped a ton! 🙏

There is an issue when parsing float numbers in custom scalars which are mapped to an `Any` type - which is the case of `Json` in your example - due to trying to read a value several times.

[A fix](https://github.com/apollographql/apollo-kotlin/pull/3839) will be in the next version, which should be released in the next few days.  In the meantime I've shown a possible workaround on your example [here](https://github.com/rev0lgenX/apollo-kotlin-sample/pull/1).

**Comment by rev0lgenX** [comment]
Thank you.

**Comment by martinbonnin** [comment]
Fixed in 3.0.1

**Comment by jasco** [comment]
Thank you. I especially appreciate the workaround patch @BoD.  Please mention this in the release notes. I earlier was hunting in release notes for any mention of parsing fixes.

**Comment by martinbonnin** [comment]
My bad, I took the changelog from the closed pull requests (like #3839 here) and not closed issues (like #3836). Definitely something we can improve for future releases. In the meantime, I've patched the 3.1.0 release notes to mention "parsing"
